solidity/contracts/bridge/VendingMachine.sol

```diff
@@ -0,0 +1,250 @@
+    function withdrawFees(address recipient, uint256 amount)
+        external
+        onlyOwner
+    {
+        tbtcV2.safeTransfer(recipient, amount);
+    }
+
+    /// @notice Allows the Governance to begin unmint fee update process.
+    ///         The update process needs to be finalized with a call to
+    ///         `finalizeUnmintFeeUpdate` function after the `GOVERNANCE_DELAY`
+    ///         passes.
+    /// @param _newUnmintFee The new unmint fee
+    function beginUnmintFeeUpdate(uint256 _newUnmintFee) external onlyOwner {
+        /* solhint-disable-next-line not-rely-on-time */
+        emit UnmintFeeUpdateStarted(_newUnmintFee, block.timestamp);
+        newUnmintFee = _newUnmintFee;
+        /* solhint-disable-next-line not-rely-on-time */
+        unmintFeeChangeInitiated = block.timestamp;
+    }
+
+    /// @notice Allows the Governance to finalize unmint fee update process.
+    ///         The update process needs to be first initiated with a call to
+    ///         `beginUnmintFeeUpdate` and the `GOVERNANCE_DELAY` needs to pass.
+    function finalizeUnmintFeeUpdate()
+        external
+        onlyOwner
+        onlyAfterGovernanceDelay(unmintFeeChangeInitiated)
+    {
+        emit UnmintFeeUpdated(newUnmintFee);
+        unmintFee = newUnmintFee;
+        newUnmintFee = 0;
+        unmintFeeChangeInitiated = 0;
+    }
+
+    /// @notice Allows the Governance to begin vending machine update process.
+    ///         The update process needs to be finalized with a call to
+    ///         `finalizeVendingMachineUpdate` function after the
+    ///         `GOVERNANCE_DELAY` passes.
+    /// @param _newVendingMachine The new vending machine address
+    function beginVendingMachineUpdate(address _newVendingMachine)
+        external
+        onlyOwner
+    {
+        require(
+            _newVendingMachine != address(0),
+            "New VendingMachine can not be zero address"
+        );
+
+        /* solhint-disable-next-line not-rely-on-time */
+        emit VendingMachineUpdateStarted(_newVendingMachine, block.timestamp);
+        newVendingMachine = _newVendingMachine;
+        /* solhint-disable-next-line not-rely-on-time */
+        vendingMachineUpdateInitiated = block.timestamp;
+    }
+
+    /// @notice Allows the Governance to finalize vending machine update process.
+    ///         The update process needs to be first initiated with a call to
+    ///         `beginVendingMachineUpdate` and the `GOVERNANCE_DELAY` needs to
+    ///         pass. Once the update is finalized, the new vending machine will
+    ///         become an owner of TBTC v2 token.
+    function finalizeVendingMachineUpdate()
+        external
+        onlyOwner
+        onlyAfterGovernanceDelay(vendingMachineUpdateInitiated)
+    {
+        emit VendingMachineUpdated(newVendingMachine);
+        //slither-disable-next-line reentrancy-no-eth
+        tbtcV2.transferOwnership(newVendingMachine);
+        newVendingMachine = address(0);
```

Comment:
can you walk me through how the vending machine update process works? We have
```
    address public newVendingMachine;
    uint256 public vendingMachineUpdateInitiated;
```
The `beginVendingMachineUpdate` method sets the new vending machine variable to the supplied address, and sets the `vendingMachineUpdateInitiated` flag.

Then, some time after the governance delay we call `finalizeVendingMachineUpdate` which clears the `vendingMachineUpdateInitiated` flag and then reverts `newVendingMachine` to the zero address.

What is actually happening here? Is the supplied address a different, updated contract and after the move the old version of the contract floats in the ether?

Reply:
>What is actually happening here? Is the supplied address a different, updated contract and after the move the old version of the contract floats in the ether?

The new VendingMachine is a new version of a contract that we will want to replace this contract.
The new VendingMachine contract will become the owner of TBTCv2 token contract which means only the new contract will be able to mint or burn TBTCv2 tokens.

Reply:
Exactly what @nkuba said. This new vending machine might be just a tweaked current implementation, for example pausing new mints for a fixed period of time so that operators of v1 have some time window to safely unwind. Unclear yet but we are giving this freedom to the DAO/community. Eventually, this will be a Vending Machine of a fully funcional tBTC v2 system.